Add Chipyard port of Zephyr RTOS

CobbledSteel opened the pull request:
<!--
First, please ensure that the title of your PR is sufficient to include in the next changelog.
Refer to https://github.com/ucb-bar/chipyard/releases for examples and feel free to ask reviewers for help.

Then, make sure to label your PR with one of the changelog:<section> labels to indicate which section
of the changelog should contain this PR's title:
  changelog:added
  changelog:changed
  changelog:fixed
  changelog:removed

If you feel that this PR should not be included in the changelog, you must still label it with
changelog:omit

Provide a brief description of the PR immediately below this comment, if the title is insufficient -->

**Summary**:
Add initial support of Zephyr RTOS to Chipyard. Intended to be an intermediate option between Linux/FireMarshal flow and BaremetalIDE flow. Evaluated on Spike and RTL sim with RocketConfig/DualRocketConfig.

**Zephyr RTOS Integration**:
- Adds Zephyr as a submodule (`chipyard/software/zephyr`)
- Implements west-based build system for Spike and Chipyard RTL simulation
- Enables Zephyr boot (~20K cycles on RocketConfig) for lightweight RTOS support

**HTIF UART Driver for Spike/FESVR**:
- Implements Zephyr UART driver for HTIF-based serial I/O
- Supports console and shell via HTIF
- Integrated into Zephyr device tree and KConfig

**Documentation Updates**:
- Adds Zephyr Getting Started guide for Chipyard
- Documents Zephyr driver structure and integration
- Provides step-by-step guide for adding new Zephyr drivers


**Related PRs / Issues**:
<!-- List any related PRs/issues here, if applicable -->

<!-- choose one -->
**Type of change**:
- [ ] Bug fix
- [x] New feature
- [ ] Other enhancement

<!-- choose one -->
**Impact**:
- [ ] RTL change
- [x] Software change (RISC-V software)
- [ ] Build system change
- [ ] Other

<!-- must be filled out completely to be considered for merging -->
**Contributor Checklist**:
- [x] Did you set `main` as the base branch?
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you state the type-of-change/impact?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you mark the PR with a `changelog:` label?
- [ ] (If applicable) Did you update the conda `.conda-lock.yml` file if you updated the conda requirements file?
- [x] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you add a test demonstrating the PR?
<!-- Do this if this PR is a bugfix that should be applied to the latest release -->
- [ ] (If applicable) Did you mark the PR as `Please Backport`?


**CI Help**:
Add the following labels to modify the CI for a set of features.
Generally, a label added only affect subsequent changes to the PR (i.e. new commits, force pushing, closing/reopening).
See `ci:*` for full list of labels:
- `ci:fpga-deploy` - Run FPGA-based E2E testing
- `ci:local-fpga-buildbitstream-deploy` - Build local FPGA bitstreams for platforms that are released
- `ci:disable` - Disable CI

jerryz123 commented on .gitmodules:
```diff
@@ -154,3 +154,6 @@
 [submodule "generators/vexiiriscv"]
 	path = generators/vexiiriscv
 	url = https://github.com/ucb-bar/vexiiriscv-tile.git
+[submodule "software/zephyr"]
```
Can you add this to .github/scripts/check-commit.sh

jerryz123 reviewed (changes requested):
If this can boot in 20k instructions, we should add a CI test that checks it using RocketConfig.

jerryz123 commented on conda-reqs/chipyard-extended.yaml:
```diff
@@ -63,6 +63,14 @@ dependencies:
+    - ccache
+    - dtc
+    - west
+    - pyelftools
```
If you update this yaml, you have to rerun the update-conda-lockfiles script.

jerryz123 reviewed (changes requested):
Zephyr is a gigantic repo, it significantly degrades clone time of chipyard. 

Can you modify scripts/init-submodules-no-riscv-tools-nolog.sh such that that script does not initialize Zephyr? You can follow the pattern of the existing repos on the "blocklist" for the recursive init.

Then indicate the user should manually `git submodule update --init zephyr` in the docs

CobbledSteel commented:
Thanks for the comments.

Modified the Zephyr submodule to no longer be cloned by default, updated the docs to reflect this.

Also added a CI test to run Zephyr hello world on RocketConfig.

jerryz123 reviewed (changes requested):
This all looks good, except one minor nit. Can the zephyr submodule be `software/zephyr`, instead of `software/zephyrproject/zephyr`?

CobbledSteel commented:
I can move it back to software/zephyr, but currently Zephyr's `west` build tool will dump several directories/config files outside of the `zephyr` directory. I moved it to `software/zephyrproject/zephyr` so that it wouldn't pollute the `software` directory.

Is generating these additional directories fine? Otherwise I'll see if it's possible to keep the submodule at `software/zephyr` and have documentation to create the `west` workspace in some clean location during setup after cloning.

jerryz123 reviewed (approved):
I see, thanks for clarifying. This is a good approach then.
